tests/test_separablePipelineExecutor.py

```diff
@@ -0,0 +1,548 @@
+
+        executor.pre_execute_qgraph(
+            graph,
+            register_dataset_types=False,
+            save_init_outputs=False,
+            save_versions=False,
+        )
+        with self.assertRaises(LookupError):
+            self.butler.datasetExists("a_config", {}, collections=[self.butler.run])
+        with self.assertRaises(LookupError):
+            self.butler.datasetExists(PipelineDatasetTypes.packagesDatasetName, {})
+
+    def test_pre_execute_qgraph_unconnected(self):
+        # Unconnected graph; see test_make_graph_nowhere_skippartial_clobber.
+        executor = SeparablePipelineExecutor(self.butler, skip_existing=True, clobber_output=True)
+        pipeline = Pipeline.fromFile(self.pipeline_file)
+        self.butler.put({"zero": 0}, "input")
+        self.butler.put({"zero": 0}, "intermediate")
+        graph = executor.make_graph(pipeline)
+
+        butlerTests.addDatasetType(self.butler, "output", set(), "StructuredDataDict")
+        butlerTests.addDatasetType(self.butler, "a_config", set(), "Config")
+        butlerTests.addDatasetType(self.butler, "b_config", set(), "Config")
+        butlerTests.addDatasetType(self.butler, "b_log", set(), "ButlerLogRecords")
+        butlerTests.addDatasetType(self.butler, "b_metadata", set(), "TaskMetadata")
+        butlerTests.addDatasetType(self.butler, PipelineDatasetTypes.packagesDatasetName, set(), "Packages")
+
+        executor.pre_execute_qgraph(
+            graph,
+            register_dataset_types=False,
+            save_init_outputs=False,
+            save_versions=False,
+        )
+        with self.assertRaises(LookupError):
+            self.butler.datasetExists("a_config", {}, collections=[self.butler.run])
+        with self.assertRaises(LookupError):
+            self.butler.datasetExists(PipelineDatasetTypes.packagesDatasetName, {})
+
+    def test_pre_execute_qgraph_empty(self):
+        executor = SeparablePipelineExecutor(self.butler)
+        graph = lsst.pipe.base.QuantumGraph({}, universe=self.butler.dimensions)
+
+        butlerTests.addDatasetType(self.butler, "output", set(), "StructuredDataDict")
+        butlerTests.addDatasetType(self.butler, "a_config", set(), "Config")
+        butlerTests.addDatasetType(self.butler, "b_config", set(), "Config")
+        butlerTests.addDatasetType(self.butler, "b_log", set(), "ButlerLogRecords")
+        butlerTests.addDatasetType(self.butler, "b_metadata", set(), "TaskMetadata")
+        butlerTests.addDatasetType(self.butler, PipelineDatasetTypes.packagesDatasetName, set(), "Packages")
+
+        executor.pre_execute_qgraph(
+            graph,
+            register_dataset_types=False,
+            save_init_outputs=False,
+            save_versions=False,
+        )
+        with self.assertRaises(LookupError):
+            self.butler.datasetExists("a_config", {}, collections=[self.butler.run])
```

Comment:
It hadn't occurred to me before that QuantumGraph doesn't allow us to distinguish between "this task is being run but has no quanta" (which logically would result in a config and other init-outputs) and "this task is not being run".  We may want to change this behavior in the future so the presence or absence of the init outputs doesn't depend on whether pre-execution was run before or after QG generation, since we won't know if there are no quanta until after QG generation.